features/config/service/impl/src/main/java/org/opennms/features/config/service/impl/AbstractCmJaxbConfigDao.java

```diff
@@ -0,0 +1,158 @@
+    private String description;
+    private final Collection<Consumer<ENTITY_CLASS>> onReloadCausedChangeCallbacks = new ArrayList<>();
+    private ConcurrentHashMap<String, ENTITY_CLASS> lastKnownEntityMap = new ConcurrentHashMap<>();
+
+    /**
+     * ConfigName use for the Config and Schema
+     *
+     * @return ConfigName
+     */
+    abstract protected String getConfigName();
+
+
+    /**
+     * The default configId when getConfig without passing configId
+     *
+     * @return configId
+     */
+    abstract protected String getDefaultConfigId();
+
+    /**
+     * <p>Constructor for AbstractJaxbConfigDao.</p>
+     *
+     * @param entityClass a {@link java.lang.Class} object.
+     * @param description a {@link java.lang.String} object.
+     */
+    public AbstractCmJaxbConfigDao(final Class<ENTITY_CLASS> entityClass, final String description) {
+        super();
+        this.entityClass = entityClass;
+        this.description = description;
+    }
+
+    /**
+     * loadConfig from CM. If the config is already stored in cache, it will also trigger reload callback
+     *
+     * @param configId
+     * @return ConfigObject
+     * @throws IOException
+     */
+    protected ENTITY_CLASS loadConfig(final String configId) throws IOException {
+        long startTime = System.currentTimeMillis();
+
+        LOG.debug("Loading {} configuration from {}", description, configId);
+        Optional<ENTITY_CLASS> configOptional = configurationManagerService.getConfiguration(this.getConfigName(), configId, entityClass);
+        if (configOptional.isEmpty()) {
+            throw new RuntimeException("NOT_FOUND: configName=" + this.getConfigName() + " configId=" + configId);
+        }
+        ENTITY_CLASS config = configOptional.get();
+        long endTime = System.currentTimeMillis();
+        LOG.info("Loaded {} in {} ms", getDescription(), (endTime - startTime));
+
+        // If the config is not load in the first time, we will trigger the callbacks for this change
+        ENTITY_CLASS lastKnownEntity = lastKnownEntityMap.get(configId);
+        if (lastKnownEntity != null) {
+            synchronized (onReloadCausedChangeCallbacks) {
+                if (!onReloadCausedChangeCallbacks.isEmpty()) {
+                    LOG.debug("Calling onReloaded callbacks");
+                    try {
+                        onReloadCausedChangeCallbacks.forEach(c -> c.accept(config));
+                    } catch (Exception e) {
+                        LOG.warn("Encountered exception while calling onReloaded callbacks", e);
+                    }
+                }
+            }
+        }
+        lastKnownEntityMap.put(configId, config);
+        return config;
+    }
+
+    /**
+     * get default config object
+     *
+     * @return config object
+     */
+    public ENTITY_CLASS getConfig(String configId) throws IOException {
+        ENTITY_CLASS lastKnownEntity = lastKnownEntityMap.get(configId);
```

Comment:
nit: could be replaced by:
return lastKnownEntityMap.computeIfAbsent(configId, this::loadConfig);

Reply:
thanks